client/inbox-panel/utils.js

```diff
@@ -52,37 +73,40 @@ export function hasValidNotes( notes ) {
 const truncateElement = ( element, limit ) => {
 	const truncatedNode = document.createElement( 'div' );
 	const childNodes = Array.from( element.childNodes );
+	const splitter = new GraphemeSplitter();
+	let truncatedTextLength = 0;
+
 	for ( let i = 0; i < childNodes.length; i++ ) {
 		// Deep clone.
 		let clone = childNodes[ i ].cloneNode( true );
-		if (
-			truncatedNode.textContent.length + clone.textContent.length <=
-			limit
-		) {
+		const cloneNodeLetters = splitter.splitGraphemes( clone.textContent );
+
+		if ( truncatedTextLength + cloneNodeLetters.length <= limit ) {
 			// No problem including a whole child node, no need to consider truncating at all.
 			truncatedNode.appendChild( clone );
+			truncatedTextLength += cloneNodeLetters.length;
+			continue;
+		}
+
+		const charactersRemaining = limit - truncatedTextLength;
+		const isNodeOnlyTextContent =
+			! clone.innerHTML ||
+			clone.textContent.slice( 0, charactersRemaining ) ===
+				clone.innerHTML.slice( 0, charactersRemaining );
```

Comment:
@chihsuan This is susceptible to miscalculation as it uses JS's native `slice` function which would not respect multi-letter characters

```js
let element = document.createElement('div')
element.textContent = '🏳️‍🌈🏳️‍🌈'
element.textContent.slice(0, 8)

> '🏳️‍🌈🏳'
```

Reply:
Good point! I used `hasChildNodes()` to check if we should call `truncateElement` instead in https://github.com/woocommerce/woocommerce-admin/pull/8404/commits/6424de37f20f2a5e3e6609d03a77832efac5bad3. 

Although it might have an additional function call compared to checking content with `charactersRemaining`. But I think this is more simple.

Reply:
Sounds good to me!